java/arcs/core/allocator/Arc.kt

```diff
@@ -91,56 +104,45 @@ class Arc internal constructor(
     /** Called whenever the [ArcState] changes to [Error]. */
     fun onError(handler: () -> Unit) = onArcStateChangeFiltered(Error, handler)
 
-    private fun fireArcStateChange() {
-        sync(this) {
-            arcStateChangeHandlers.forEach {
-                it(arcState)
-            }
-        }
+    private fun recomputeArcState(states: Collection<ArcState>): ArcState = when {
+        states.any { it == Deleted } -> Deleted
+        states.any { it == Error } -> Error
+        states.all { it == Running } -> Running
+        states.all { it == Stopped } -> Stopped
+        states.all { it == NeverStarted } -> NeverStarted
+        else -> Indeterminate
     }
 
-    private fun recomputeArcState() {
-        sync(this) {
-            val states = arcStatesByHostId.values
-            arcState = when {
-                states.any { it == Deleted } -> Deleted
-                states.any { it == Error } -> Error
-                states.all { it == Running } -> Running
-                states.all { it == Stopped } -> Stopped
-                states.all { it == NeverStarted } -> NeverStarted
-                else -> Indeterminate
-            }
+    private fun maybeRegisterChangeHandlerWithArcHosts() {
+        if (!registered.compareAndSet(false, true)) {
+            return
         }
-    }
-
-    private suspend fun fetchCurrentStates() {
-        partitions.forEach {
-            val arcHost = allocator.lookupArcHost(it.arcHost)
-            arcStatesByHostId[it.arcHost] = arcHost.lookupArcHostStatus(it)
-        }
-    }
 
-    private fun registerChangeHandlerWithArcHosts() = sync(this) {
-        require(registration == null) {
-            "registration called more than once"
-        }
-        runBlocking {
-            // first poll the current states of all hosts
-            fetchCurrentStates()
-
-            // Register event listeners
+        arcStatesByHostFlow = callbackFlow {
             partitions.forEach { partition ->
                 val arcHost = allocator.lookupArcHost(partition.arcHost)
-                registration = arcHost.addOnArcStateChange(id) { _, state ->
-                    sync(this) {
-                        arcStatesByHostId[partition.arcHost] = state
-                        recomputeArcState()
-                    }
+                registrations[partition.arcHost] = arcHost.addOnArcStateChange(id) { _, state ->
+                    offer(partition.arcHost to state)
```

Comment:
Will the `offer` occur after the `closeFlow`? Perhaps check `isClosedForSend`.

Reply:
close() is implemented by enqueuing a Close operation on the flow. I looked at AbstractChannel and they seem to try and account for this by scanning for any pending stuff queued after the close token and draining it (https://github.com/Kotlin/kotlinx.coroutines/blob/583ec6e862fb70c7fc0232af948a028bab3b20de/kotlinx-coroutines-core/common/src/channels/AbstractChannel.kt#L274)

However, in the interests of safety in case their logic is buggy or doesn't do what I think, I'll add a isClosedForSend check to be safe.